provider/equinix/environ.go

```diff
@@ -0,0 +1,665 @@
+	var subnetIDs []string
+	for _, subnetList := range args.SubnetsToZones {
+		for subnetID := range subnetList {
+			packetSubnetID := strings.TrimPrefix(subnetID.String(), "subnet-")
+			subnetIDs = append(subnetIDs, packetSubnetID)
+		}
+	}
+
+	return subnetIDs, nil
+}
+
+// supportedInstanceTypes returns the instance types supported by Equnix Metal.
+func (e *environ) supportedInstanceTypes() ([]instances.InstanceType, error) {
+	opt := &packngo.ListOptions{
+		Includes: []string{"available_in"},
+	}
+	plans, _, err := e.equnixClient.Plans.List(opt)
+	if err != nil {
+		return nil, errors.Annotate(err, "retrieving supported instance types")
+	}
+
+	var instTypes []instances.InstanceType
+nextPlan:
+	for _, plan := range plans {
+		if !validPlan(plan, e.cloud.Region) {
+			logger.Infof("Plan %s not valid in facility %s", plan.Name, e.cloud.Region)
+			continue
+		}
+
+		var instArch string
+		switch {
+		case strings.HasSuffix(plan.Name, ".x86"):
+			instArch = arch.AMD64
+		case strings.HasSuffix(plan.Name, ".arm"):
+			instArch = arch.ARM64
+		default:
+			continue nextPlan
+
+		}
+
+		mem, err := parseMemValue(plan.Specs.Memory.Total)
+		if err != nil {
+			continue
+		}
+
+		instTypes = append(instTypes,
+			instances.InstanceType{
+				Id:       plan.ID,
+				Name:     plan.Name,
+				CpuCores: uint64(plan.Specs.Cpus[0].Count),
+				Mem:      mem,
+				Arches:   []string{instArch},
+				// Scale per hour costs so they can be represented as an integer for sorting purposes.
+				Cost: uint64(plan.Pricing.Hour * 1000.0),
+				// TODO: returned by packet's API but not exposed by the packngo client
+				// Deprecated: plan.Legacy,
+			})
+	}
+
+	return instTypes, nil
+}
+
+func validPlan(plan packngo.Plan, region string) bool {
+	notAvailable := true
+	for _, a := range plan.AvailableIn {
+		if a.Code == region {
+			notAvailable = false
+			break
+		}
+	}
+	isInvalid := notAvailable || plan.Pricing == nil ||
+		plan.Specs == nil ||
+		plan.Specs.Memory == nil ||
+		len(plan.Specs.Cpus) == 0 || plan.Specs.Cpus[0].Count == 0
+
+	return !isInvalid
+}
+
+func parseMemValue(v string) (uint64, error) {
+	var scaler = uint64(1)
+	if strings.HasSuffix(v, "GB") {
+		scaler = 1024
+		v = strings.TrimSuffix(v, "GB")
+	}
+
+	val, err := strconv.ParseUint(v, 10, 64)
+	return val * scaler, err
+}
+
+func (e *environ) findInstanceSpec(controller bool, allImages []*imagemetadata.ImageMetadata, instanceTypes []instances.InstanceType, ic *instances.InstanceConstraint) (*instances.InstanceSpec, error) {
+	oss, _, err := e.equnixClient.OperatingSystems.List()
+	if err != nil {
+		return nil, err
+	}
+	suitableImages := []*imagemetadata.ImageMetadata{}
+
+	for _, it := range instanceTypes {
+		for _, os := range oss {
+			if !isDistroSupported(os, ic) {
+				continue
+			}
+
+			for _, p := range os.ProvisionableOn {
+				if p == it.Name {
+					image := &imagemetadata.ImageMetadata{
+						Id:   os.Slug,
+						Arch: arch.AMD64,
```

Comment:
What about other arches? L478 indicates that arm64 is also supported.

Reply:
What do you suggest we do about it?

Reply:
Can we glean any arch-related information from `p` here? If not, and if packet guarantees that you can deploy distros to arm64 machines, you could probably generate two entries here? (Ideally, to future-proof the implementation, you could make the supported arches as a package-level variable and iterate it here creating an ImageMetadata entry for each supported arch.)

Reply:
we can't get much from `p` this is all the information I get from packngo.OS 
```
type OS struct {
	Name            string   `json:"name"`
	Slug            string   `json:"slug"`
	Distro          string   `json:"distro"`
	Version         string   `json:"version"`
	ProvisionableOn []string `json:"provisionable_on"`
}
``` 
Where `ProvisionableOn` is just an array of string with plan names  "c1.small.x86",  "baremetal_1",  "c1.xlarge.x86",
Maybe if we were to extract that data from plan which has property
```
 "cpus": [
        {
          "count": 1,
          "type": "Intel(R) Xeon(R) CPU E3-1578L v5 @ 2.00GHz"
        }
      ],
``` 
which we would have to normalize to extract correct architecture

Reply:
Ok, so AFAICT, in one of the other methods we extract the arch from the plan name. You could extract that code into a helper that given the plan name, it returns an `arch.Arch` value and use the result (if any) to populate the `Arch` field in the generated `ImageMetadata` value. Would that do the trick?

Reply:
would this work
```
func getArchitectureFromPlan(p string) string {
	planSplit := strings.Split(p, ".")
	var architecture string
	if len(planSplit) > 3 {
		architecture = planSplit[2]
	}
	switch architecture {
	case "x86":
		return arch.AMD64
	case "arm":
		return arch.ARM64
	default:
		return arch.AMD64
	}
}
```